SSA/Core/MLIRSyntax/Parser.lean

```diff
@@ -0,0 +1,138 @@
+      | .error e s => .error s!" Failed elaborating {stx}.\nUnable to pretty-print exception at 'elabIntoEIO':\n{e}." s
+
+def region0Expr := (Expr.app (Expr.const `MLIR.AST.Region []) (Expr.const `Nat.zero []))
+def op0Expr := (Expr.app (Expr.const `MLIR.AST.Op []) (Expr.const `Nat.zero []))
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Region`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabRegion (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Region 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Region 0) env region0Expr stx
+  return reg
+
+/--
+  Parse `Lean.Syntax` into `MLIR.AST.Op`.
+
+  This runs the Lean frontend stack and adds it to the TCB.
+  It should be safe (if we trust the lean parser) since we ensure that the types match: see
+  https://leanprover-community.github.io/mathlib4_docs/Std/Util/TermUnsafe.html#Std.TermUnsafe.termUnsafe_
+ -/
+def elabOp (env : Lean.Environment) (stx : Lean.Syntax) : EIO ParseError (MLIR.AST.Op 0) := do
+  let reg ← unsafe elabIntoEIO (α := MLIR.AST.Op 0) env op0Expr stx
+  return reg
+
+def mkParseFun (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (s : String) (env : Environment) : EIO ParseError ParseOutput := do
+  let syn : Except String Syntax := Parser.runParserCategory env syntaxcat s
+  match syn with
+    | .error msg => throw msg
+    | .ok syn => ntparser syn
+
+private def mkNonTerminalParser (syntaxcat : Name)
+    (ntparser : Syntax → EIO ParseError ParseOutput)
+    (env : Environment) (s : String) : EIO ParseError ParseOutput :=
+  let parseFun := mkParseFun syntaxcat ntparser
+  parseFun s env
+
+-- TODO: do we need this copy of the environment, or can I remove it from one of the two?
+def regionParser : @ParseFun (MLIR.AST.Region 0) :=
+  fun env : Lean.Environment => mkNonTerminalParser `mlir_region (elabRegion env) env
+
+private def parseFile (env: Lean.Environment)
+    (parser: @ParseFun ParseOutput)
+    (filepath: System.FilePath) : IO (Option ParseOutput) := do
+  let lines ← IO.FS.lines filepath
+  let fileStr := lines.foldl (λ s₁ s₂ => s₁ ++ "\n" ++ s₂) "" -- TODO: use intercalate.
+  let parsed ← EIO.toIO' <| parser env fileStr
+  match parsed with
+  | .ok parseOutput => return parseOutput
+  | .error msg => IO.println s!"Error parsing {filepath}:\n{msg}"; return none
+
+private def isFile (p: System.FilePath) : IO Bool := do
+  return (← p.metadata).type == IO.FS.FileType.file
+
+def runParser (parser : @ParseFun ParseOutput) (fileName : String) : IO (Option ParseOutput) := do
+  let executablePath : System.FilePath ← IO.appPath
+  let dotLakePath : System.FilePath :=
+    match executablePath.parent with
```

Comment:
This feels very fragile to me; Can we generalize this, to walk up the parent tree till it finds a `.lake` folder, throwing an error if it doesn't find one, rather than assuming the depth?

Reply:
@bollu agreed! However, I' d feel that suggestion is also too fragile, because it will also not work if we at some point have a ` make install`. For that reason I' m not happy with that change, I'd request that we either do it right right (i.e. by having a configurable prefix or an $SSA_HOME or something) or we leave it like it is until we have a proper solution.

Reply:
@goens how about making the user give us the root repo directory, and we navigate relative to this?

Reply:
with a default argument? yeah that sounds like a good compromise!

Reply:
> with a default argument? yeah that sounds like a good compromise!

Yes, where the default argument, IMO, recursively walks the directory tree to find a `.lake`.

Reply:
I'm trying to avoid writing code that recusively walks down the directory tree to find a `.lake` because it feels like not so useful work (if the directory structure changes it breaks), I can find almost no use-case for it over what we just have + being able to provide that directory explicitly